modules/core/src/mathfuncs.cpp

```diff
@@ -1600,50 +1608,311 @@ static bool ocl_patchNaNs( InputOutputArray _a, float value )
+                v_int32 sh2 = v_rotate_right<1>(vand);
+                ve = (v_reinterpret_as_u64(vand) & mask10) | (v_reinterpret_as_u64(sh2) & ~mask10);
+            }
+            #endif
+
+            vv[j] = ve;
+        }
+
+        v_uint8 v = v_pack_b(vv[0], vv[1], vv[2], vv[3], z, z, z, z);
+
+        v_store_low(dst + i, v);
+    }
+
+    return i;
+}
+
+#endif
+
+
+template <typename _Tp, int cn>
+static void finiteMask_(const _Tp *src, uchar *dst, size_t total)
+{
+    size_t i = 0;
+
+#if CV_SIMD
+    i = finiteMaskSIMD_<_Tp, cn>(src, dst, total);
 #endif
 
-        for( ; j < len; j++ )
-            if( (tptr[j] & 0x7fffffff) > 0x7f800000 )
-                tptr[j] = val.i;
+    for(; i < total; i++ )
+    {
+        bool nan = true;
+        for (int c = 0; c < cn; c++)
+        {
+            _Tp val = src[i * cn + c];
+            nan = nan && !cvIsNaN(val) && !cvIsInf(val);
+        }
+        dst[i] = nan ? 255 : 0;
     }
 }
 
+#ifdef HAVE_OPENCL
+
+
+static bool ocl_finiteMask(const UMat img, UMat mask)
+{
+    int channels = img.channels();
+    int depth = img.depth();
+    int rowsPerWI = ocl::Device::getDefault().isIntel() ? 4 : 1;
+    ocl::Kernel k("finiteMask", ocl::core::finitemask_oclsrc,
+                  format("-D srcT=%s -D cn=%d -D rowsPerWI=%d",
+                         depth == CV_32F ? "float" : "double", channels, rowsPerWI));
+    if (k.empty())
+        return false;
+
+    k.args(ocl::KernelArg::ReadOnlyNoSize(img), ocl::KernelArg::WriteOnly(mask));
+
+    size_t globalsize[2] = { (size_t)img.cols, ((size_t)img.rows + rowsPerWI - 1) / rowsPerWI };
+    return k.run(2, globalsize, NULL, false);
 }
 
+#endif
+
+void finiteMask(InputArray _img, OutputArray _mask)
+{
+    CV_INSTRUMENT_REGION();
+
+    int channels = _img.channels();
+    int depth = _img.depth();
+    CV_Assert( depth == CV_32F || depth == CV_64F );
+    std::vector<int> vsz(_img.dims());
+    _img.sizend(vsz.data());
+    _mask.create(_img.dims(), vsz.data(), CV_8UC1);
+
+    CV_OCL_RUN(_img.isUMat() && _mask.isUMat() && _img.dims() <= 2,
+               ocl_finiteMask(_img.getUMat(), _mask.getUMat()));
+
+    Mat img = _img.getMat();
+    Mat mask = _mask.getMat();
+
+    const Mat *arrays[]={&img, &mask, 0};
+    Mat planes[2];
+    NAryMatIterator it(arrays, planes);
+    size_t total = planes[0].total();
+    size_t i, nplanes = it.nplanes;
+
+    for( i = 0; i < nplanes; i++, ++it )
+    {
+        const uchar* sptr = planes[0].ptr();
+        uchar* dptr = planes[1].ptr();
+
+        switch( depth )
+        {
+        case CV_32F:
+            switch (channels)
+            {
+            case 1: finiteMask_<float, 1>((const float*)sptr, dptr, total); break;
+            case 2: finiteMask_<float, 2>((const float*)sptr, dptr, total); break;
+            case 3: finiteMask_<float, 3>((const float*)sptr, dptr, total); break;
+            case 4: finiteMask_<float, 4>((const float*)sptr, dptr, total); break;
+            }
```

Comment:
`channels` values are not validated at all.
For `channels=5` function does nothing and doesn't throw any exception.

Reply:
Fixed